test/pangolin-exchange.test.ts

```diff
@@ -0,0 +1,104 @@
+]
+
+const pangolinRouterAbi = [
+  'function getAmountsIn (uint256 amountOut, address[] path) view returns (uint256[])'
+]
+
+
+describe('PangolinExchange', () => {
+  describe('Test buying and selling an asset', () => {
+    let sut: PangolinExchange,
+      daie: Contract,
+      pgrt: Contract,
+      owner: SignerWithAddress;
+
+    before('Deploy the PangolinExchange contract', async () => {
+      [owner] = await getFixedGasSigners(10000000);
+      sut = await deployContract(owner, PangolinExchangeArtifact, [pangolinRouterAddress]) as PangolinExchange;
+      daie = await new ethers.Contract(daiETokenAddress, ERC20Abi);
+      pgrt = await new ethers.Contract(pangolinRouterAddress, pangolinRouterAbi);
+    });
+
+
+    it('should check for the amount of tokens to buy to be greater than 0', async () => {
+      await expect(sut.buyERC20Token(daiETokenAddress, 0)).to.be.revertedWith('Amount of tokens to buy has to be greater than 0');
+    });
+
+
+    it('should check if enough funds were provided', async () => {
+      const daiePurchaseAmount = 1e18;
+      const estimatedAvax = (await pgrt.connect(owner).getAmountsIn(daiePurchaseAmount.toString(), [WAVAXTokenAddress, daiETokenAddress]))[0];
+
+      await expect(sut.buyERC20Token(daiETokenAddress, daiePurchaseAmount.toString(), {value: Math.floor(estimatedAvax*0.9).toString()})).to.be.revertedWith('Not enough funds provided');
+    });
+
+
+    it('should check if an erc20 tokens were purchased successfully', async () => {
+      const DAIePurchaseAmount = 1e18;
+      const estimatedAvax = (await pgrt.connect(owner).getAmountsIn(DAIePurchaseAmount.toString(), [WAVAXTokenAddress, daiETokenAddress]))[0];
+      const initialDAIeBalance = await daie.connect(owner).balanceOf(owner.address);
+      const initialAvaxBalance = await provider.getBalance(owner.address);
+
+      expect(initialDAIeBalance).to.equal(0);
+
+      await sut.buyERC20Token(daiETokenAddress, DAIePurchaseAmount.toString(), {value: estimatedAvax.toString()});
+
+      const currentDAIeBalance = await daie.connect(owner).balanceOf(owner.address);
+      const currentAvaxBalance = await provider.getBalance(owner.address);
+      const avaxBalanceDifference = BigNumber.from(initialAvaxBalance.toString()).sub(BigNumber.from(estimatedAvax.toString()));
+
+      expect(currentDAIeBalance).to.equal(DAIePurchaseAmount.toString());
+      expect(currentAvaxBalance).to.be.lt(avaxBalanceDifference);
+    });
+
+
+    it('should check for the amount of tokens to sell to be greater than 0', async () => {
+      await expect(sut.sellERC20Token(daiETokenAddress, 0)).to.be.revertedWith('Amount of tokens to sell has to be greater than 0');
+    });
+
+
+    it('should check for a sufficient token allowance', async () => {
+      await expect(sut.sellERC20Token(daiETokenAddress, 1)).to.be.revertedWith('Insufficient token allowance');
+    });
+
+
+    it('should check if an erc20 tokens were sold successfully', async () => {
```

Comment:
I'll also add a test case when we sell part of the tokens owned.

Reply:
That is exactly what this test does - selling half (1e18/2) of the tokens owned.